packages/langium/src/documents/document-builder.ts

```diff
@@ -55,6 +59,32 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
         };
     }
 
+    validate(document: LangiumDocument): Diagnostic[] | undefined {
+        let diagnostics: Diagnostic[] | undefined;
+        const validator = this.documentValidator;
+        if (this.connection) {
```

Comment:
This method is used only in `build`, which already checks whether `connection` is available. Do we need this method at all? We could do the same within `build` as before.

Reply:
If I don't check this.connection I will need to do `!.`: `this.connection!.sendDiagnostics({ uri: document.textDocument.uri, diagnostics });`

> Do we need this method at all? 

I think yes. First, to easily override only this aspect of building a document. Second, I am still not sure that validation should be a part of document building. For not opened documents I wold like to do only parsing and linking for example. I think we can target that refactoring in the next PR's, what do you think?

Reply:
Yes, the default (but overridable) behavior should be to validate only open documents, which can be checked with the TextDocuments service. Could you create an issue so we don't forget?

Reply:
I'd propose to make the method protected and remove it from the interface.